collector/collector.go

```diff
@@ -400,14 +400,26 @@ func fetchRemoteConfig(url string) (*Config, error) {
 }
 
 func HasValidHeader(data string) bool {
-	header := strings.SplitN(data, "\n", 2)[0]
-
-	// Trim trailing whitespaces
-	header = strings.TrimRightFunc(header, unicode.IsSpace)
+	// Get the first 10 lines
+	headers := strings.SplitN(data, "\n", 10)
+
+	// iterate over them as there could be comments or the jinja template info:
+	// https://cloudinit.readthedocs.io/en/latest/explanation/instancedata.html#example-cloud-config-with-instance-data
+
+	for _, line := range headers {
+		// Trim trailing whitespaces
+		header := strings.TrimRightFunc(line, unicode.IsSpace)
+		// If it starts with a hash check it, in case its a huge line, we dont want to waste time
+		if strings.HasPrefix(header, "#") {
```

Comment:
This allows things like this:

```
# Some comment
users:
  - name: whatever

#cloud-config

other: true
```

So, as long as `#cloud-config` is in the first 10 lines, this code is ok with it. I think we need to define the rule better and be a bit more strict.

I can't find any docs on how many comment lines cloud-init allows before the header but I'm thinking we could keep reading lines until:

- a line that doesn't start with `#` is encountered. This means there is no header and we return false.
- a valid header is encountered: we return true
- end of data: we return false

Reply:
Well mainly to be a bit flexible but not that much.

So we allow some flexibility on the header but don't read huge files and lose time on it.

Basically extend the header stuff as a courtesy for comments or that Jinja stuff, but not overcomplicate it, if your header is not on the first ten lines then bad luck.

I don't think it needs that much of work for this to be honest.

Reply:
ok let's keep it simple for now. If there is need we can revisit this.